### ISSUE TITLE
Android Integration

### ISSUE THREAD

**Issue by adiviness** [ColoradoSchoolOfMines/interface_sdk#19]
We should include android smartphones as a sensor device. I want to be able to use accelerometer/gyro/etc data from my phone to control visualizations. Or beam data to the phone to be kept 'secret' from the other players. Think like battleship or something. I don't know, I just want phone support. Maybe something like this too:

https://www.photobeamer.com/

**Comment by muff1nman** [comment]
Agreed. 

Some questions:
- Where would it fit in architecturally? 
- Would we have a "server" of a type in the hardware layer and people could scan a qr code to get an app to use their phone as a client device?  
- Would it then use bluetooth?  
- Can we stream multiple bluetooth connections or would we have to use something like wifi?

**Comment by adiviness** [comment]
It would have to be a separate hardware device.

There would probably have to be some sort of app they would have to
download in order to communicate, I'm not sure how it would work though
considering that phones would be much more ephemeral than a kinect, for
instance. Some sort of server interface to connect to would probably be
easiest.

I'm not very familiar with bluetooth, I can't see why it wouldn't work
though, unless whatever receiver had a constricting limit on the number of
devices that can connect at once.

**Comment by muff1nman** [comment]
Yeah. I would agree with the app thing.  That would be a good way to abstract out the differences in android phones as well and possibly even ios.  (who cares about blackberries lol)  I guess I ask about the bluetooth just because I have never seen two devices connected to a single device at the same time...

**Comment by adiviness** [comment]
http://stackoverflow.com/questions/16266079/connecting-multiple-devices-to-bluetooth-in-android

It sounds like you can. While I'm thinking about how to get them to
download the app, a qr code would work well, however I think an nfc tag
glued somewhere nearby with a qr code printed on top of it (for good
measure) is even better.

**Comment by muff1nman** [comment]
Now were talking!  haha

The one thing that would be interesting is how the bandwidth / latency of bluetooth would affect it.  I mean we would have to be smart about if say we had a sub hardware device / functionality to get acceleration from a phone that we don't overload the phone by requesting its acceleration too often.  imagine if we did this every frame in a processing app... ouch.  The other way to go about it is to only make it so that the phone pushes its info to the server.  Then this hardware device would be strictly accessed via events.  The nice thing about this is that we could make this a per user / phone app setting so that phones that are more sluggish could have their update frequencies turned down in case their phone starts acting up...

**Comment by adiviness** [comment]
I do agree that it becomes more of a problem when we want to target a type
of device that has wildly varying hardware specs. I don't know the answer
to the problem, unfortunately.

**Comment by muff1nman** [comment]
Some possibly relevant resources:

[link](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&cad=rja&ved=0CCsQFjAA&url=http%3A%2F%2Ftoday.java.net%2Fpub%2Fa%2Ftoday%2F2004%2F07%2F27%2Fbluetooth.html&ei=e3HTUde2Ler1ygH9uYHgDQ&usg=AFQjCNHT9Kpxn6Xmz_e0kUsKvniZtnX7KQ&sig2=p40003gZZO1CpwUMY4zrVQ)
[link](https://code.google.com/p/bluecove/)

**Comment by muff1nman** [comment]
OpenNI2 supports android for the record. Not sure exactly what we could do with it though...

**Comment by adiviness** [comment]
All sorts of stuff. For example: Racing game? Tilt phone to accelerate and
turn.
